File: java/tst/TestCompoundIndex.java
import org.garret.perst.*;

import java.util.Iterator;

public class TestCompoundIndex { 
    final static int nRecords = 100000;
    final static int pagePoolSize = 32*1024*1024;

    static class Record extends Persistent { 
        String strKey;
        int    intKey;
    };

    static public void main(String[] args) {    
        Storage db = StorageFactory.getInstance().createStorage();

        for (int i = 0; i < args.length; i++) { 
            if ("altbtree".equals(args[i])) { 
                db.setProperty("perst.alternative.btree", Boolean.TRUE);
            }
        }
        db.open("testcidx.dbs", pagePoolSize);
        FieldIndex root = (FieldIndex)db.getRoot();
        if (root == null) { 
            root = db.createFieldIndex(Record.class, new String[]{"intKey", "strKey"}, true);
            db.setRoot(root);
        }
        long start = System.currentTimeMillis();
        long key = 1999;
        int i;
        for (i = 0; i < nRecords; i++) { 
            Record rec = new Record();
            key = (3141592621L*key + 2718281829L) % 1000000007L;
            rec.intKey = (int)(key >>> 32);
            rec.strKey = Integer.toString((int)key);
            root.put(rec);                
        }
        db.commit();
        System.out.println("Elapsed time for inserting " + nRecords + " records: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");
        
        start = System.currentTimeMillis();
        key = 1999;
        int minKey = Integer.MAX_VALUE;
        int maxKey = Integer.MIN_VALUE;
        for (i = 0; i < nRecords; i++) { 
            key = (3141592621L*key + 2718281829L) % 1000000007L;
            int intKey = (int)(key >>> 32);            
            String strKey = Integer.toString((int)key);
            Record rec = (Record)root.get(new Key(new Object[]{new Integer(intKey), strKey}));
            Assert.that(rec != null && rec.intKey == intKey && rec.strKey.equals(strKey));
            if (intKey < minKey) { 
                minKey = intKey;
            }
            if (intKey > maxKey) { 
                maxKey = intKey;
            }
        }
        System.out.println("Elapsed time for performing " + nRecords + " index searches: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");
        
        start = System.currentTimeMillis();
        Iterator iterator = root.iterator(new Key(new Integer(minKey), ""), 
                                          new Key(new Integer(maxKey+1), "???"), 
                                          FieldIndex.ASCENT_ORDER);
        int n = 0;
        String prevStr = "";
        int prevInt = minKey;
        while (iterator.hasNext()) { 
            Record rec = (Record)iterator.next();
            Assert.that(rec.intKey > prevInt || rec.intKey == prevInt && rec.strKey.compareTo(prevStr) > 0);
            prevStr = rec.strKey;
            prevInt = rec.intKey;
            n += 1;
        }
        Assert.that(n == nRecords);
        
        iterator = root.iterator(new Key(new Integer(minKey), "", false), 
                                 new Key(new Integer(maxKey+1), "???", false), 
                                 FieldIndex.DESCENT_ORDER);
        n = 0;
        prevInt = maxKey+1;
        while (iterator.hasNext()) { 
            Record rec = (Record)iterator.next();
            Assert.that(rec.intKey < prevInt || rec.intKey == prevInt && rec.strKey.compareTo(prevStr) < 0);
            prevStr = rec.strKey;
            prevInt = rec.intKey;
            n += 1;
        }
        Assert.that(n == nRecords);
        System.out.println("Elapsed time for iterating through " + (nRecords*2) + " records: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");
        start = System.currentTimeMillis();
        key = 1999;
        for (i = 0; i < nRecords; i++) { 
            key = (3141592621L*key + 2718281829L) % 1000000007L;
            int intKey = (int)(key >>> 32);            
            String strKey = Integer.toString((int)key);
            Record rec = (Record)root.get(new Key(new Object[]{new Integer(intKey), strKey}));
            Assert.that(rec != null && rec.intKey == intKey && rec.strKey.equals(strKey));
            Assert.that(root.contains(rec));
            root.remove(rec);
            rec.deallocate();
        }
        Assert.that(!root.iterator().hasNext());
        Assert.that(!root.iterator(null, null, Index.DESCENT_ORDER).hasNext());
        Assert.that(!root.iterator(null, null, Index.ASCENT_ORDER).hasNext());
        System.out.println("Elapsed time for deleting " + nRecords + " records: " 
                           + (System.currentTimeMillis() - start) + " milliseconds");
        db.close();
    }
}

File: csharp/tests/Tests/TestCompoundIndex.cs
namespace Volante
{
    using System;

    public class TestCompoundResult : TestResult
    {
        public TimeSpan InsertTime;
        public TimeSpan IndexSearchTime;
        public TimeSpan IterationTime;
        public TimeSpan RemoveTime;
    }

    public class TestCompoundIndex : ITest
    {
        class Record : Persistent
        {
            internal String strKey;
            internal int intKey;
        }

        public void Run(TestConfig config)
        {
            int i;
            int count = config.Count;
            var res = new TestCompoundResult();
            config.Result = res;

            DateTime start = DateTime.Now;

            IDatabase db = config.GetDatabase();
            IMultiFieldIndex<Record> root = (IMultiFieldIndex<Record>)db.Root;
            Tests.Assert(root == null);
            root = db.CreateFieldIndex<Record>(new string[] { "intKey", "strKey" }, IndexType.Unique);
            db.Root = root;

            long key = 1999;
            for (i = 0; i < count; i++)
            {
                Record rec = new Record();
                rec.intKey = (int)((ulong)key >> 32);
                rec.strKey = Convert.ToString((int)key);
                root.Put(rec);
                key = (3141592621L * key + 2718281829L) % 1000000007L;
            }
            db.Commit();
            res.InsertTime = DateTime.Now - start;

            start = DateTime.Now;
            key = 1999;
            int minKey = Int32.MaxValue;
            int maxKey = Int32.MinValue;
            for (i = 0; i < count; i++)
            {
                int intKey = (int)((ulong)key >> 32);
                String strKey = Convert.ToString((int)key);
                Record rec = root.Get(new Key(new Object[] { intKey, strKey }));
                Tests.Assert(rec != null && rec.intKey == intKey && rec.strKey.Equals(strKey));
                if (intKey < minKey)
                    minKey = intKey;
                if (intKey > maxKey)
                    maxKey = intKey;
                key = (3141592621L * key + 2718281829L) % 1000000007L;
            }
            res.IndexSearchTime = DateTime.Now - start;

            start = DateTime.Now;
            int n = 0;
            string prevStr = "";
            int prevInt = minKey;
            foreach (Record rec in root.Range(new Key(minKey, ""),
                                              new Key(maxKey + 1, "???"),
                                              IterationOrder.AscentOrder))
            {
                Tests.Assert(rec.intKey > prevInt || rec.intKey == prevInt && rec.strKey.CompareTo(prevStr) > 0);
                prevStr = rec.strKey;
                prevInt = rec.intKey;
                n += 1;
            }
            Tests.Assert(n == count);

            n = 0;
            prevInt = maxKey + 1;
            foreach (Record rec in root.Range(new Key(minKey, "", false),
                                              new Key(maxKey + 1, "???", false),
                                              IterationOrder.DescentOrder))
            {
                Tests.Assert(rec.intKey < prevInt || rec.intKey == prevInt && rec.strKey.CompareTo(prevStr) < 0);
                prevStr = rec.strKey;
                prevInt = rec.intKey;
                n += 1;
            }
            Tests.Assert(n == count);
            res.IterationTime = DateTime.Now - start;
            start = DateTime.Now;
            key = 1999;
            for (i = 0; i < count; i++)
            {
                int intKey = (int)((ulong)key >> 32);
                String strKey = Convert.ToString((int)key);
                Record rec = root.Get(new Key(new Object[] { intKey, strKey }));
                Tests.Assert(rec != null && rec.intKey == intKey && rec.strKey.Equals(strKey));
                Tests.Assert(root.Contains(rec));
                root.Remove(rec);
                rec.Deallocate();
                key = (3141592621L * key + 2718281829L) % 1000000007L;
            }
            Tests.Assert(!root.GetEnumerator().MoveNext());
            Tests.Assert(!root.Reverse().GetEnumerator().MoveNext());
            res.RemoveTime = DateTime.Now - start;
            db.Close();
        }
    }

}
